### PR TITLE
Add geographiclib/1.51

### PR DESCRIPTION
**Pull request by hdhauk**
- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by SpaceIm** [comment]
You have also to raise ConanInvalidConfiguration if version >= 1.51 and clang < 4 I guess.

**Comment by hdhauk** [comment]
> You have also to raise ConanInvalidConfiguration if version >= 1.51 and clang < 4 I guess.

I've added a `tools.check_min_cppstd(self, 11)`, wouldn't that suffice?

**Comment by SpaceIm** [comment]
No, because it prevents to build with compilers versions for which default C++ standard is not 11 or higher, like gcc 4.9 or 5.

**Comment by dmn-star** [comment]
You can adopt _min_compiler_version_default_cxx11 from the boost recipe
https://github.com/conan-io/conan-center-index/pull/3977/files/15e9b1be20a61892d5958ed494ab74a9d7b7b333

File: recipes/geographiclib/all/conanfile.py [hunk 1]
```diff
@@ -23,9 +24,24 @@ def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
 
+    @property
+    def _min_compiler_version_default_cxx11(self):
+        # Minimum compiler version having c++ standard >= 11
+        return {
+            "gcc": 6,
+            "clang": 6,
+            "Visual Studio": 14,  # guess
```
Comment (SpaceIm) [inline recipes/geographiclib/all/conanfile.py:33]:
```suggestion
            "gcc": 4.9,
            "clang": 6,
            "Visual Studio": 14,  # guess
```

I'm able to build with gcc 4.9 and 5

File: recipes/geographiclib/all/conanfile.py [hunk 2]
```diff
@@ -23,9 +24,24 @@ def config_options(self):
...
+        }.get(str(self.settings.compiler))
```
Comment (SpaceIm) [inline recipes/geographiclib/all/conanfile.py:34]:
```suggestion
        }.get(str(self.settings.compiler), False)
```

File: recipes/geographiclib/all/conanfile.py [hunk 3]
```diff
@@ -23,9 +24,24 @@ def config_options(self):
...
+
     def configure(self):
         if self.options.shared:
             del self.options.fPIC
+        if tools.Version(self.version) >= "1.51":
+            if self.settings.compiler.cppstd:
+                tools.min_cppstd(self, 11)
+            elif tools.Version(self.settings.compiler.version) < self._min_compiler_version_default_cxx11:
+                raise ConanInvalidConfiguration("C++11 support needed for version >= 1.51")
```
Comment (SpaceIm) [inline recipes/geographiclib/all/conanfile.py:43]:
```suggestion
            minimum_version = self._min_compiler_version_default_cxx11
            if not minimum_version:
                self.output.warn("geographiclib {} requires C++11 math functions. Your compiler is unknown. Assuming it supports this feature.".format(self.version))
            elif tools.Version(self.settings.compiler.version) < minimum_version:
                raise ConanInvalidConfiguration("geographiclib {} requires C++11 math functions, which your compiler does not support.".format(self.version))
```

**Review by SpaceIm — COMMENTED** [review]
Don't raise for compilers you have not listed in `_min_compiler_version_default_cxx11()`

File: recipes/geographiclib/all/conanfile.py (hunk 1 above)
Comment (dmn-star) [reply recipes/geographiclib/all/conanfile.py:33]: "gcc": 4.9 -> "gcc": 6.1
The default mode is C++98 for GCC versions prior to 6.1, and C++14 for GCC 6.1 and above.

Comment (SpaceIm) [reply recipes/geographiclib/all/conanfile.py:33]: It supports C++11 even if it's not the default.

Comment (dmn-star) [reply recipes/geographiclib/all/conanfile.py]: it doesn't matter, the important thing is that it has only c++03 default mode.

And  c++11 support in gcc 4.9 is quite buggy. And anyway, what's the point of using  almost EOL compiler?

1. Use compiler.cppstd if set (  On CCI cppstd is not set anywhere)
2. Use C++ standard compiler _default mode_, if cppstd not set

Comment (SpaceIm) [reply recipes/geographiclib/all/conanfile.py]: Most underlying build files of open source libraries don't follow default standard, they force any standard allowing to properly build if user didn't ask for a specific one.
So if cppstd is not set, I think that it should not be default standard of the compiler, but any standard supported by the compiler and allowing to build the library.

Comment (dmn-star) [reply recipes/geographiclib/all/conanfile.py]: Please read the discussion #3977. :-)

Comment (SpaceIm) [reply recipes/geographiclib/all/conanfile.py]: It's a different issue. Like most libraries, geographiclib doesn't conditionnally disable build of components depending on C++ standard, boost is an edge case.

Comment (dmn-star) [reply recipes/geographiclib/all/conanfile.py]: >It's a different issue. 
for the same reason. apple-clang (with c++17 support) was used in C++03 default mode to build the boost.

>conditionnally disable build of components depending on C++ standard, boost is an edge case.
but what about the  conditional compilation depending on C++ standard or other other c++ funny things?

Why not follow the official Conan guide , bump gcc to 6.1 and profit. :-)

Comment (SpaceIm) [reply recipes/geographiclib/all/conanfile.py]: There is no official Conan guide about cppstd AFAIK, and cppstd in profile is experimental.

Visual Studio don't have C++17 as default standard for example. Do you want to not build pre build binaries at all for Visual Studio if library requires C++17?

Comment (dmn-star) [reply recipes/geographiclib/all/conanfile.py]: VS2019 IDE pass the c++14 flag to cl.exe by default.   :-) Xcode 12 does too for apple-clang **but** CCI uses apple-clang in C++03 mode on the command line. 

>Do you want to not provide or test in CCI binaries for Visual Studio if library requires C++17?
Why not, if the underlying build system do right thing ( cmake for example target_compile_features(xxxx _PUBLIC_ cxx_std_17)?  But not by force cppstd in the recipe.

>There is no official Conan guide about cppstd AFAIK, and cppstd in profile is experimental.
IMHO https://docs.conan.io/en/latest/howtos/manage_cpp_standard.html 

No matter simply never use the binaries generated by Conan in the production.

**Comment by hdhauk** [comment]
> Failure in build 7 (`239e31bcfe803d6d6a51c0ef523e352dfeb236f4`):
> 
>     * **Error processing recipe (ref 'geographiclib/1.50.1'): Linux x86_64, Release, gcc 4.9, libstdc++ . Options:  geographiclib:shared-True**
>       Unexpected error happened:
> 
> 
> ```
> ERROR: 400: Unsupported Conan v2 repository request for 'c3i_PR-4291_239e31bcfe803d6d6a51c0ef523e352dfeb236f4'. [Remote: c3i_PR-4291_239e31bcfe803d6d6a51c0ef523e352dfeb236f4]
> ```

Does anyone know what might have gone wrong here? Clicking the details link gives me a 403 error. @SpaceIm @dmn-star

**Comment by dmn-star** [comment]
try to close the PR and reopen it after ~20--30 sec.

File: recipes/geographiclib/all/conanfile.py (hunk 1 above)
Comment (SpaceIm) [reply recipes/geographiclib/all/conanfile.py]: > > Do you want to not provide or test in CCI binaries for Visual Studio if library requires C++17?
> 
> Why not, if the underlying build system do right thing ( cmake for example target_compile_features(xxxx _PUBLIC_ cxx_std_17)? But not by force cppstd in the recipe.

Yes and it's what most recipes do, they rely on underlying build system to set any minimum standard, so they don't honor default standard of compiler.

File: recipes/geographiclib/all/conanfile.py [hunk 4]
```diff
@@ -35,8 +54,16 @@ def source(self):
...
+            if tools.Version(self.version) >= "1.51":
+                if self.settings.compiler.get_safe("cppstd"):
+                    tools.check_min_cppstd(self, 11)
+                elif self._compiler_supports_cxx11:
+                    self._cmake.definitions["CONAN_CMAKE_CXX_STANDARD"] = 11
+                else:
+                    raise ConanInvalidConfiguration("C++11 support needed for version >= 1.51")
+
+        self._cmake.definitions["GEOGRAPHICLIB_LIB_TYPE"] = "SHARED" if self.options.shared else "STATIC"
+        self._cmake.configure(build_folder=self._build_subfolder)
```
[the remark this answers is not in the document]
Comment (hdhauk) [reply recipes/geographiclib/all/conanfile.py:66]: Oops! I'll fix that indentation error. Am I missing something tho? Wouldn't this achieve what we want? The first case will make sure we get a ConanInvalidConfiguration exception if `settings.compiler.cppstd` is set too low, and the second will make sure we use C++11  if it isn't set and the compiler supports it.

[the remark this answers is not in the document]
Comment (SpaceIm) [reply recipes/geographiclib/all/conanfile.py]: It's already handled in underlying build files. Your errors come from test_package, you need to set CXX_STANDARD 11.

I've proposed modifications far above, you wan still use them ;)

File: recipes/geographiclib/all/test_package/conanfile.py [hunk 5]
```diff
@@ -8,6 +8,7 @@ class TestPackageConan(ConanFile):
     def build(self):
         cmake = CMake(self)
         cmake.configure()
+        cmake.definitions["CONAN_CMAKE_CXX_STANDARD"] = 11
```
Comment (SpaceIm) [inline recipes/geographiclib/all/test_package/conanfile.py:11]:
```suggestion
```

and add those lines in CMakeLists.txt of test_package

```cmake
if(geographiclib_VERSION VERSION_GREATER_EQUAL "1.51")
  set_property(TARGET ${PROJECT_NAME} PROPERTY CXX_STANDARD 11)
endif()
```

File: recipes/geographiclib/all/conanfile.py [hunk 6]
```diff
@@ -35,9 +54,17 @@ def source(self):
...
+                    tools.check_min_cppstd(self, 11)
+                elif self._compiler_supports_cxx11:
+                    self._cmake.definitions["CONAN_CMAKE_CXX_STANDARD"] = 11
+                else:
+                    raise ConanInvalidConfiguration("C++11 support needed for version >= 1.51")
+
             self._cmake.definitions["GEOGRAPHICLIB_LIB_TYPE"] = "SHARED" if self.options.shared else "STATIC"
             self._cmake.configure(build_folder=self._build_subfolder)
-        return self._cmake
+            return self._cmake
```
Comment (SpaceIm) [inline recipes/geographiclib/all/conanfile.py:67]:
```suggestion
        return self._cmake
```

undo modifications

File: recipes/geographiclib/all/conanfile.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hdhauk) [reply recipes/geographiclib/all/conanfile.py]: Alright! Thanks :) I'll try to make those changes then

**Review by SpaceIm — APPROVED** [review]

File: recipes/geographiclib/all/conanfile.py [hunk 7]
```diff
@@ -23,9 +24,34 @@ def config_options(self):
...
+        }.get(str(self.settings.compiler), False)
+
     def configure(self):
         if self.options.shared:
             del self.options.fPIC
+        if tools.Version(self.version) >= "1.51":
+            if self.settings.compiler.cppstd:
+                tools.min_cppstd(self, 11)
+
+            def lazy_lt_semver(v1, v2):
```
Comment (prince-chrismc) [inline recipes/geographiclib/all/conanfile.py:44]: Another variant https://github.com/conan-io/conan/issues/8002 (just to keep track of all of them)

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**